### PR TITLE
Hot reload accelerated table on dataset update

### LINKED ISSUE #1084: Unloading and then reloading datasets shouldn't clear out data
**Describe the bug**
When a dataset is hot reloaded in the runtime, it is first unregistered from DataFusion and then the new dataset is registered. This can lead to a period of time where the new dataset hasn't finished initializing/loading data and it appears there are 0 rows.

We should wait for the new dataset to finish its initialization before unloading the old dataset and replacing it.

### PR DESCRIPTION
**Pull request by ewgenius**
closes #1084

Refactored dataset update logic:
For accelerated datasets (only memory mode), spice runtime performs hot reload:

- create new accelerated table and wait until it loaded with data, while keeping old dataset registered in datafusion
- once new table successfully loaded, unregister old dataset and register new one
- in case of failure it falls back to regular reload

https://github.com/spiceai/spiceai/assets/827338/5e3838cb-4be4-44e7-82b7-90ae930fc551

### CODE REVIEW

**Review by digadeesh — DISMISSED** [review]

File: crates/runtime/src/accelerated_table.rs [hunk 1]
```diff
@@ -68,7 +68,7 @@ pub type Result<T> = std::result::Result<T, Error>;
 // An accelerated table consists of a federated table and a local accelerator.
 //
 // The accelerator must support inserts.
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs:69]: Let's update the comment to indicate what the oneshot channel is doing.

File: crates/runtime/src/datafusion.rs [hunk 2]
```diff
@@ -145,8 +174,19 @@ impl DataFusion {
         &mut self,
         dataset: impl Borrow<Dataset>,
         table: Table,
+        accelerated_table: Option<AcceleratedTable>,
```
Comment (phillipleblanc) [inline crates/runtime/src/datafusion.rs]: This should be a new enum for `Table`

File: crates/runtime/src/datafusion/schema.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use datafusion::{
+    catalog::schema::SchemaProvider,
+    datasource::TableProvider,
+    error::{DataFusionError, Result},
+};
+
+// Copy of default MemorySchemaProvider, with allowed table register overwrites
```
Comment (phillipleblanc) [inline crates/runtime/src/datafusion/schema.rs]:
```suggestion
// Copy of default MemorySchemaProvider that allows `register_table` to atomically overwrite any existing tables
```

File: crates/runtime/src/lib.rs [hunk 4]
```diff
@@ -372,11 +385,31 @@ impl Runtime {
     }
 
     pub async fn update_dataset(&self, ds: &Dataset, all_datasets: &[Dataset]) {
-        tracing::info!("Updating dataset: {}...", &ds.name);
         status::update_dataset(&ds.name, status::ComponentStatus::Refreshing);
         if let Ok(connector) = self.load_dataset_connector(ds, all_datasets).await {
+            if ds.is_file_accelerated() {
+                tracing::warn!("File accelerated datasets doesn't support hot reload, falling back to full dataset reload");
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]:
```suggestion
                tracing::warn!("File accelerated datasets currently don't support hot reload, falling back to full dataset reload");
```

File: crates/runtime/src/datafusion.rs (hunk 2 above)
Comment (ewgenius) [reply crates/runtime/src/datafusion.rs]: thanks, updated

File: crates/runtime/src/accelerated_table.rs (hunk 1 above)
Comment (ewgenius) [reply crates/runtime/src/accelerated_table.rs:69]: Added a comment

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**